cpp/CMakeLists.txt

```diff
@@ -415,226 +412,45 @@ endif(CONDA_LINK_DIRS)
-            src/strings/replace/replace_re.cu
-            src/strings/replace/backref_re.cu
-            src/strings/replace/backref_re_medium.cu
-            src/strings/replace/backref_re_large.cu
-            src/strings/replace/multi_re.cu
-            src/strings/replace/replace.cu
-            src/strings/sorting/sorting.cu
-            src/strings/split/partition.cu
-            src/strings/split/split.cu
-            src/strings/split/split_record.cu
-            src/strings/strings_column_factories.cu
-            src/strings/strings_column_view.cu
-            src/strings/strings_scalar_factories.cpp
-            src/strings/strip.cu
-            src/strings/substring.cu
-            src/strings/translate.cu
-            src/strings/utilities.cu
-            src/lists/extract.cu
-            src/lists/lists_column_factories.cu
-            src/lists/lists_column_view.cu
-            src/lists/copying/concatenate.cu
-            src/lists/copying/gather.cu
-            src/lists/copying/copying.cu
-            src/structs/structs_column_view.cu
-            src/structs/structs_column_factories.cu
-            src/text/detokenize.cu
-            src/text/edit_distance.cu
-            src/text/generate_ngrams.cu
-            src/text/normalize.cu
-            src/text/stemmer.cu
-            src/text/tokenize.cu
-            src/text/ngrams_tokenize.cu
-            src/text/replace.cu
-            src/text/subword/load_hash_file.cu
-            src/text/subword/data_normalizer.cu
-            src/text/subword/wordpiece_tokenizer.cu
-            src/text/subword/subword_tokenize.cu
-            src/scalar/scalar.cpp
-            src/scalar/scalar_factories.cpp
-            src/dictionary/add_keys.cu
-            src/dictionary/detail/concatenate.cu
-            src/dictionary/dictionary_column_view.cpp
-            src/dictionary/dictionary_factories.cu
-            src/dictionary/decode.cu
-            src/dictionary/encode.cu
-            src/dictionary/remove_keys.cu
-            src/dictionary/search.cu
-            src/dictionary/set_keys.cu
-            src/groupby/groupby.cu
-            src/groupby/hash/groupby.cu
-            src/groupby/sort/groupby.cu
-            src/groupby/sort/sort_helper.cu
-            src/groupby/sort/group_sum.cu
-            src/groupby/sort/group_min.cu
-            src/groupby/sort/group_max.cu
-            src/groupby/sort/group_argmax.cu
-            src/groupby/sort/group_argmin.cu
-            src/groupby/sort/group_count.cu
-            src/groupby/sort/group_nunique.cu
-            src/groupby/sort/group_nth_element.cu
-            src/groupby/sort/group_std.cu
-            src/groupby/sort/group_quantiles.cu
-            src/aggregation/aggregation.cpp
-            src/aggregation/aggregation.cu
-            src/aggregation/result_cache.cpp
-)
-
-# Override RPATH for cudf
-set_target_properties(cudf PROPERTIES BUILD_RPATH "\$ORIGIN")
+function(add_library_module MODULE MODULE_DIR)
```

Comment:
I'm not familiar with library modules. Can you clarify the question? What are the options?

Reply:
> I'm not familiar with library modules. Can you clarify the question? What are the options?

So each module is a candidate for either a statically compiled archive (`libcudf_agg.a`) or a shared object (`libcudf_agg.so`).

My concern is that if we have many shared objects that we're incurring a performance cost in symbol lookup or function calls (analogous to what we found with RDC).

Currently, I plan to run the benchmark suite to clarify what the performance of this solution is, but if there are obvious gotcha's, I'd love to know if I've missed something.

Reply:
I'm looking for feedback on where we should place the the library partition boundaries.
For example, all files under `src/aggregation` are considered a part of the `cudf_agg` module.
This results in 33 modules based on directory structure. I picked this split to get something out the door.

Is there feedback you have regarding how this split should be done? Programmatically, it is simple to put the split anywhere; so, it seems appropriate to place a split where we won't have to touch it again.

Consider an abstract `cudf_base` module: does it make sense to lump `comms copying interop column table structs stream_compaction io hash bitmask` together and call that base?

Reply:
Our docs do some grouping.
![image](https://user-images.githubusercontent.com/783069/91942461-551de800-ed3e-11ea-941b-75c032695227.png)

I collapsed some groups to try to come up with some high-level grouping...

Reply:
I would say:
cudf_core = column and table, utilities, bitmask operations, anything type related (not algorithmic), hashing, interop
cudf_copying (include reshaping in here?)
cudf_sorting (include sorting, merging, and reordering)
cudf_searching
cudf_joining
cudf_aggregation (include quantiles here?)
cudf_transformation
cudf_io
cudf_strings (include nvtext)

If any of these are too big, there are opportunities to split.

Reply:
Two updates.
1. Given the volume of modules that `libcudf` is being broken into, I believe it makes sense to introduce a `cudf::` namespace (e.g.) instead of referring to `cudf_core`, users will link against `cudf::core` in CMake.
2. The library modules are somewhat tangled (codependent). `cudf_base` (what we're calling `cudf_core`) requires more than just types + containers.

Regarding `cudf_copying`, I worry about adding _anything_ to that module. It is by far the largest module. As the library continues to grow, `cudf_copying` presents the highest risk of exceeding the maximum allowable size for debugging symbols. That said, I can put `reshaping` into `copying` and we can just split it later on if it becomes necessary.